boot_shutdown_interface/src/boot_shutdown_interface.cpp

```diff
@@ -0,0 +1,154 @@
+: Node("boot_shutdown", options),
+  ecu_name_(declare_parameter("ecu_name", std::string("ecu"))),
+  startup_timeout_(declare_parameter("startup_timeout", 300)),
+  preparation_timeout_(declare_parameter("preparation_timeout", 60)),
+  // Artificial delay for debugging
+  preparation_delay_(declare_parameter("debug.preparation_delay", 5))
+{
+  using namespace std::literals::chrono_literals;
+  using std::placeholders::_1;
+  using std::placeholders::_2;
+
+  ecu_state_.name = ecu_name_;
+  ecu_state_.state = EcuState::STARTUP;
+  startup_time_ = get_clock()->now();
+
+  srv_prepare_shutdown_ = create_service<PrepareShutdown>(
+    "~/service/prepare_shutdown", std::bind(
+      &BootShutdownInterface::onPrepareShutdown, this, _1,
+      _2),
+    rmw_qos_profile_default);
+
+  srv_execute_shutdown_ = create_service<ExecuteShutdown>(
+    "~/service/execute_shutdown", std::bind(
+      &BootShutdownInterface::onExecuteShutdown, this, _1,
+      _2),
+    rmw_qos_profile_default);
+
+  pub_ecu_state_ =
+    create_publisher<EcuState>("~/output/ecu_state", rclcpp::QoS{1});
+
+  auto qos = rclcpp::QoS{1};
+  if (declare_parameter(
+      "topic_config.transient_local",
+      static_cast<rclcpp::ParameterValue>(false)).get<bool>())
+  {
+    qos.transient_local();
+  }
+  if (declare_parameter(
+      "topic_config.best_effort",
+      static_cast<rclcpp::ParameterValue>(false)).get<bool>())
+  {
+    qos.best_effort();
+  }
+
+  sub_topic_ = create_generic_subscription(
+    declare_parameter("topic_config.name", rclcpp::PARAMETER_STRING).get<std::string>(),
+    declare_parameter("topic_config.type", rclcpp::PARAMETER_STRING).get<std::string>(),
+    qos, std::bind(&BootShutdownInterface::onTopic, this, std::placeholders::_1));
+
+  timer_ = rclcpp::create_timer(
+    this, get_clock(), 1s, std::bind(&BootShutdownInterface::onTimer, this));
+}
+
+void BootShutdownInterface::onPrepareShutdown(
+  PrepareShutdown::Request::SharedPtr, PrepareShutdown::Response::SharedPtr response)
+{
+  RCLCPP_INFO(get_logger(), "shutdown preparing...");
+
+  // Clear pagecache, dentries, and inodes.
+  preparation_child_ = boost::process::child(
+    "/bin/sh", "-c",
+    fmt::format("sync; echo 3 > /proc/sys/vm/drop_caches; sleep {}", preparation_delay_));
+
+  ecu_state_.state = EcuState::SHUTDOWN_PREPARING;
+  preparation_start_time_ = get_clock()->now();
+
+  response->status.success = true;
+  response->power_off_time =
+    builtin_interfaces::msg::Time(get_clock()->now() + rclcpp::Duration(preparation_delay_, 0));
+}
+
+void BootShutdownInterface::onExecuteShutdown(
+  ExecuteShutdown::Request::SharedPtr, ExecuteShutdown::Response::SharedPtr response)
+{
+  RCLCPP_INFO(get_logger(), "shutdown executing...");
+
+  // Shutdown
+  // Avoid using "now" and delay shutdown to publish ecu_state
+  shutdown_child_ = boost::process::child("/bin/sh", "-c", fmt::format("shutdown -h now"));
+
+  auto power_off_time = builtin_interfaces::msg::Time(get_clock()->now());
```

Comment:
No need to use `builtin_interfaces::msg::Time`.

Reply:
The `power_off_time` should be an estimated time to execute shutdown instead of current time like https://github.com/tier4/boot_shutdown_tools/blob/ce2bde3f76a57eaee95e61b90558333255507498/boot_shutdown_interface/src/boot_shutdown_guest.cpp#L73

Reply:
@ito-san 
I removed use of `builtin_interfaces::msg::Time` and 
changed the `power_off_time` to estimated shutdown execution time.
6516b0ef9b117751ab3df4840f489ee60d2d56df